osu.Game.Rulesets.Osu/Scoring/OsuHealthProcessor.cs

```diff
@@ -1,25 +1,80 @@
 // Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
 // See the LICENCE file in the repository root for full licence text.
 
+using System;
 using osu.Game.Beatmaps;
 using osu.Game.Rulesets.Judgements;
-using osu.Game.Rulesets.Objects;
 using osu.Game.Rulesets.Objects.Types;
+using osu.Game.Rulesets.Osu.Judgements;
 using osu.Game.Rulesets.Osu.Objects;
 using osu.Game.Rulesets.Scoring;
 
 namespace osu.Game.Rulesets.Osu.Scoring
 {
     public partial class OsuHealthProcessor : DrainingHealthProcessor
     {
+        private ComboResult currentComboResult = ComboResult.Perfect;
+
         public OsuHealthProcessor(double drainStartTime, double drainLenience = 0)
             : base(drainStartTime, drainLenience)
         {
         }
 
-        protected override int? GetDensityGroup(HitObject hitObject) => (hitObject as IHasComboInformation)?.ComboIndex;
-
         protected override double GetHealthIncreaseFor(JudgementResult result)
+        {
+            if (IsSimulating)
+                return getHealthIncreaseFor(result);
+
+            if (result.HitObject is not IHasComboInformation combo)
+                return getHealthIncreaseFor(result);
```

Comment:
I guess this is for safety but this branch should be 100% dead...? `OsuHitObject` implements `IHasComboInformation`.

Reply:
I suppose so. Imo there's nothing wrong with being safe. Might eventually move this to the base implementation if it works well too.